cozy/codegen/java.py

```diff
@@ -649,7 +649,7 @@ def array_in_bounds(self, elem_type, a, i):
         else:
             len = EEscape("{a}.length", ["a"], [a]).with_type(INT)
         return EAll([
-            EBinOp(i, ">=", ZERO).with_type(BOOL),
+            EBinOp(EBinOp(i, "+", ONE).with_type(INT), ">=", ZERO).with_type(BOOL),
```

Comment:
in the caller side, I passed in `s.capacity - 1` rather than `s.capacity` to `array_resize_for_index`, because I think index is size - 1 semantically. However, when capacity/size is zero, then index is -1, which in my mind is a special index, which array_in_bounds is always true. Does this make sense to you?

Another approach is to add a zero-valued-capacity guard inside the "EnsureCapacity" routine,  like `if capacity > 0, then resize(capacity - 1), else do nothing`. However, the code generated doesn't work...

Reply:
The intended purpose of `array_resize_for_index` is that it grows the array `a` until `a[i]` will execute without exception if `i >= 0`.

The change you made to `visit_SEnsureCapacity` therefore looks safe since the array has capacity `c` if `a[c-1]` will execute without exception.

The intended meaning of `SEnsureCapacity` with a run time capacity value 0 is that the array grows until it contains at least zero elements---which any array satisfies. It will call `array_resize_for_index` with a run time value `i = -1`. I expect the code should then do nothing. Without your fix I guess it loops infinitely.

I think all we need to do is add the condition `i >= 0` to the while-condition on line 627 in `array_resize_for_index`. Then add docstrings explaining what these functions do so that future readers can better understand what this code does.

Reply:
Nice point. fixed!